service/worker/scheduler/workflow.go

```diff
@@ -0,0 +1,800 @@
+		s.logger.Debug("watcher returned for running workflow")
+		return
+	}
+
+	// now we know it's not running, remove from running workflow list
+	match := func(ex *commonpb.WorkflowExecution) bool { return ex.WorkflowId == id }
+	if idx := slices.IndexFunc(s.Info.RunningWorkflows, match); idx >= 0 {
+		s.Info.RunningWorkflows = slices.Delete(s.Info.RunningWorkflows, idx, idx+1)
+	} else {
+		s.logger.Error("closed workflow not found in running list", "workflow", id)
+	}
+
+	// handle pause-on-failure
+	failedStatus := res.Status == enumspb.WORKFLOW_EXECUTION_STATUS_FAILED || res.Status == enumspb.WORKFLOW_EXECUTION_STATUS_TIMED_OUT
+	pauseOnFailure := s.Schedule.Policies.PauseOnFailure && failedStatus && !s.Schedule.State.Paused
+	if pauseOnFailure {
+		s.Schedule.State.Paused = true
+		if res.Status == enumspb.WORKFLOW_EXECUTION_STATUS_FAILED {
+			s.Schedule.State.Notes = fmt.Sprintf("paused due to workflow failure: %s: %s", id, res.GetFailure().GetMessage())
+			s.logger.Info("paused due to workflow failure", "workflow", id, "message", res.GetFailure().GetMessage())
+		} else if res.Status == enumspb.WORKFLOW_EXECUTION_STATUS_TIMED_OUT {
+			s.Schedule.State.Notes = fmt.Sprintf("paused due to workflow timeout: %s", id)
+			s.logger.Info("paused due to workflow timeout", "workflow", id)
+		}
+		s.incSeqNo()
+	}
+
+	// handle last completion/failure
+	if res.GetResult() != nil {
+		s.State.LastCompletionResult = res.GetResult()
+		s.State.ContinuedFailure = nil
+	} else if res.GetFailure() != nil {
+		// leave LastCompletionResult from previous run
+		s.State.ContinuedFailure = res.GetFailure()
+	}
+
+	s.logger.Info("started workflow finished", "workflow", id, "status", res.Status, "pause-after-failure", pauseOnFailure)
+}
+
+func (s *scheduler) handleUpdateSignal(ch workflow.ReceiveChannel, _ bool) {
+	var req schedspb.FullUpdateRequest
+	ch.Receive(s.ctx, &req)
+	if err := s.checkConflict(req.ConflictToken); err != nil {
+		s.logger.Warn("Update conflicted with concurrent change")
+		return
+	}
+
+	s.logger.Info("Schedule update", "new-schedule", req.Schedule.String())
+
+	s.Schedule.Spec = req.Schedule.Spec
+	s.Schedule.Action = req.Schedule.Action
+	s.Schedule.Policies = req.Schedule.Policies
+	s.Schedule.State = req.Schedule.State
+	// don't touch Info
+
+	s.ensureFields()
+	s.compileSpec()
+
+	s.Info.UpdateTime = timestamp.TimePtr(s.now())
+	s.incSeqNo()
+}
+
+func (s *scheduler) handlePatchSignal(ch workflow.ReceiveChannel, _ bool) {
+	var patch schedpb.SchedulePatch
+	ch.Receive(s.ctx, &patch)
+	s.logger.Info("Schedule patch", "patch", patch.String())
+	s.processPatch(&patch)
+}
+
+func (s *scheduler) handleRefreshSignal(ch workflow.ReceiveChannel, _ bool) {
+	var refresh schedspb.RefreshRequest
+	ch.Receive(s.ctx, &refresh)
+	s.logger.Debug("Got refresh signal", "refresh", refresh.String())
+	// TODO: consider just setting needRefresh to true so we refresh all instead of sending a list
```

Comment:
thoughts on this one, or getting rid of needRefresh entirely and always refreshing status?

Reply:
Should not require caller who sends signals to know those workflowIds (which might require to query first). 
And yes, I think the total number of running workflows should be small for majority of the case, and so having one flag to refresh all make sense.

Reply:
the caller is Describe: it does need to query first and then check the status of all the workflows, because we don't want every single Describe to result in a refresh signal (which goes in history), only the ones where we actually need to refresh. but I can simplify it.